feat: 게시글 업로드 기능 구현

Hansangjin98 opened the pull request:
❗️ 이슈 번호
Closes #15 

### 📝 작업 유형

- [x] 신규 기능 추가

### 📙 작업 내역

- 커스텀 메뉴 버튼, 카테고리 스택 뷰 구현
- Lottie 기반 로딩 뷰 구현
- 게시글 업로드 View, Reactor 구현
- 게시글 업로드 Domain, Service 구현 및 서버 API 연동
- FileManager 싱글톤 구조체 구현

Hansangjin98 committed:
feat: #15 - DesignKit TextField&Alert 코드 확장 및 수정, 이미지 에셋 추가

Hansangjin98 committed:
feat: #15 - 재사용 가능한 뷰 및 열거형 구현

- 메뉴 버튼 구현
- 카테고리 스택뷰 구현
- 네트워크 에러 열거형 구현

Hansangjin98 committed:
feat: #15 - 게시글 업로드 Service 레이어 구현

- API Response 모델 구현
- Service 인터페이스 및 구현체 구현
- API 연동

Hansangjin98 committed:
feat: #15 - 게시글 업로드 Domain 레이어 구현, coordinatorAssembly 코드 수정

Hansangjin98 committed:
feat: #15 - Lottie 파일 추가 및 로딩 뷰 구현

- Lottie 파일 추가
- Tuist 의존성 추가
- 커스텀 로딩 뷰 구현

Hansangjin98 committed:
feat: #15 - FileManager 구조체 구현, Domain 및 Service 일부 수정

- Core 모듈에 FileManager 구조체 구현
- VideoFile Entity 구현
- Entity 구현에 따른 Domain 및 Service 파라미터 수정
- FileManager 활용을 위한 Info.plist 수정

Hansangjin98 committed:
feat: #15 - 게시글 업로드 Feature 레이어 구현, 오타 수정

- 게시글 업로드 View, Reactor 구현
- TabBar 게시글 업로드 View 연동
- Home Reactor 오타 수정
- Coordinator, SceneFactory에 게시글 업로드View 연동

Hansangjin98 committed:
feat: #15 - AlertViewController 바인딩 수정

Hansangjin98 committed:
style: #15 - 네이밍 수정, 공백 제거

nil-ian commented on Targets/Core/Sources/PhoChakFileManager.swift:
```diff
@@ -0,0 +1,61 @@
+
+import Foundation
+
+public struct PhoChakFileManager {
```
1. `struct` 로 정의한 이유 궁금!
2. Upload 쪽에서만 활용될 것 같은데 싱글톤으로 생성한 이유가 있을까?
3. 지금 구조에서는 업로드 관련 로직에서 이 싱글톤 객체랑 강하게 결합이 되어있어서 테스트가 힘들 것 같아~ DIP 준수해서 인터페이스 통해 주입받는 형태로 제공하면 어떨까?

Hansangjin98 replied:
다른 뷰에서도 사용될 수 있을 것 같아서 싱글톤 구조체로 구현했었는데 enum + static 함수로 변경했어!!

View, Reactor, Service에서 모두 파일매니저에 접근해야돼서 인터페이스를 통해 주입받는 것 보다 static 함수로 전역에서 접근할 수 있도록 하는 게 깔끔해 보였는데 혹시 이런 경우도 테스트가 힘든 경우일까? 잘 몰라서..😭

nil-ian replied:
지금 코드는 Reactor, Service 등 비즈니스 로직을 처리하는 객체들을 단위테스트 할 때에도 파일매니저 접근을 통해 실질적인 I/O가 발생하게 되는 형태인데 이걸 인터페이스 기반으로 주입받으면 위 일련의 과정을 수행한 척 하는 목 객체를 활용해서 테스트에 수월하지 않을까 싶은뎁!?
참고만 해줘~

Hansangjin98 replied:
이해 완료 + 반영 완료 !!!

nil-ian commented on Targets/Core/Sources/PhoChakFileManager.swift:
```diff
@@ -0,0 +1,61 @@
+  public static let shared = PhoChakFileManager()
+  private let fileManager = FileManager.default
+
+  private var documentDirectoryURL: URL {
```
이 변수는 아래에 `updatedVideoFolderURL` 을 만들때만 접근하는 것으로 보이는데 별도 변수 저장 목적이 없다면 제거하는게 어떨까?

nil-ian commented on Targets/Core/Sources/PhoChakFileManager.swift:
```diff
@@ -0,0 +1,61 @@
+          attributes: nil
+        )
+      } catch {
+        print("⚠️ \(#function) - \(#line)")
```
save 실패시에도 유저는 성공한것으로 알 게 될 것 같은데 save, fetch, remove에 대한 에러 발생시 어떻게 처리할지 기획 회의해보는게 좋을것 같아~

Hansangjin98 replied:
이건 조금만 더 고민해볼게!!

nil-ian commented on Targets/Core/Sources/PhoChakFileManager.swift:
```diff
@@ -0,0 +1,61 @@
+
+    let path = uploadedVideosFolderURL.appendingPathComponent(name)
+
+    DispatchQueue.global(qos: .userInteractive).async {
```
qos 인자에 userInteractive를 전달한 기준이 있을까~?

Hansangjin98 replied:
유저 로컬 디렉토리에 빠르게 데이터를 복사해야해서 습관적으로 넣었는데 필요 없을 것 같아서 제거했엉!!

nil-ian replied:
아~ 지금 플로우에서는 업로드 할 영상을 선택하고 일부 내용을 기입한 이후에 업로드를 요청하니까 기입할 내용이 많지 않은 경우에는 상황에 따라 높은 우선순위가 필요할수도 있겠다~

nil-ian commented on Targets/Core/Sources/PhoChakFileManager.swift:
```diff
@@ -0,0 +1,61 @@
+
+  // MARK: Methods
+  public func saveVideo(name: String, data: Data) {
+    if !fileManager.fileExists(atPath: uploadedVideosFolderURL.path) {
```
내가 파일매니저 쪽을 잘 몰라서ㅠㅠ 접근하고자 하는 디렉토리에 이미 파일이 있을 경우에 대한 에러처리나 방어코드는 없는데 그럴 가능성은 없을까?

Hansangjin98 replied:
fileExists 메서드로 '동영상을 담을 폴더가 존재하는지'를 체크하고 있는 건데
폴더가 존재하지 않으면 폴더를 만들어주고, 없는 경우에는 해당 폴더에 동영상 파일을 담는 형식이라서 괜찮을 것 같아!

nil-ian replied:
간단히 코드만 봤을땐 PickerView에서 영상을 선택하면 `saveVideo`를 호출해서 파일매니저 I/O가 실행되고 영상 업로드를 완료하면 modal을 dismiss 하면서 `removeUploadedVideos` 를 통해 기존에 I/O 를 원복하는거로 확인했는데 만일 PickerView에서 영상 선택 이후 앱을 강제종료 했다가 다시 동일 스텝 밟게 되면 이미 file이 exist 상태여서 영상 업로드가 안될것 같은데 이 케이스에 대한 테스트도 되었을까~?

Hansangjin98 replied:
PickerView에서 선택할 때 마다 같은 영상이더라도 파일명은 난수로 생성되는거라 아마 곂칠 일은 없을 것 같은데 혹시 모르니까 방어코드 추가해놓을게!!

nil-ian commented on Targets/Core/Sources/PhoChakFileManager.swift:
```diff
@@ -0,0 +1,61 @@
+    return "\(uploadedVideosFolderURL.path)/\(name)"
+  }
+
+  public func removeUploadedVideos() {
```
메소드명이 복수형으로 네이밍 되어있다ㅎㅎ

Hansangjin98 replied:
유저가 동영상을 선택했다가 재선택 버튼으로 다시 선택하게 되면 로컬에 이전에 선택한 영상+새로 선택한 영상이 같이 담기게 되는데
업로드가 완료되면 담겼던 동영상들을 전부 삭제시키는 메서드라서 복수형으로 네이밍 했어!

nil-ian replied:
재선택한 영상과 기존 선택의 영상이 다르다면  videoFileName을 기반으로 path를 생성하다보니 두 Video가 저장된 path가 각각 다를것 같은데 아래 코드로 두 개 모두 삭제가 되는거야~?

Hansangjin98 replied:
영상들은 전부 한 폴더 내부에 저장되는데, 영상들이 모여있는 그 폴더 자체를 삭제해버리는 거라서 다같이 삭제돼!

nil-ian commented on Targets/DesignKit/Sources/Views/PhoChakAnimationView.swift:
```diff
@@ -0,0 +1,75 @@
+import SnapKit
+import Then
+
+public final class PhoChakAnimationView: UIViewController {
```
1. ViewController를 상속받으니 suffix는 일관성잇게 ViewController 가는게 좋을듯!
2. 로딩시에만 사용되는 ViewController면 `PhochakLoadingViewController` 라는 이름으로 네이밍하면 더 직관적일것 같은데 어떄?

nil-ian commented on Targets/DesignKit/Sources/Views/PhoChakAnimationView.swift:
```diff
@@ -0,0 +1,75 @@
+
+  // MARK: Methods
+  public func play(on parentViewController: UIViewController) {
+    parentViewController.present(self, animated: true)
```
1. 로딩VC는 코디네이터 통해서 띄우지 않는 이유가 있을까?
2. 지금 동작에서는 present, hide 할 떄 그라데이션이 깔린 배경이 올라왔다가 내려왔다가 하면서 로딩뷰 느낌으로는 약간 부자연스러울것 같아~ 관련해서 성인이에게 물어보니까 present,hide시 alpha값 애니메이션 줘서 일반적인 로딩 팝업 띄우는 느낌으로 띄우는게 좋을것 같다네!

Hansangjin98 replied:
로딩VC가 present된 이후에 특별히 수행하는 로직 없이 잠깐 나왔다가 사라지는 VC이기도 하고 View에서 play메서드를 통해 present하는 게 개인적으로 좀 더 깔끔해보여서 코디네이터를 스킵했어!

2번은 반영 완료~!

nil-ian commented on Targets/Domain/Sources/UseCases/UploadVideoPostUseCase.swift:
```diff
@@ -0,0 +1,63 @@
+
+import RxSwift
+
+public enum UploadVideoPostEvent {
```
비디오 업로드 성공/실패에 대한 열거형이면 `UploadVideoPostResult` 이런 형태로 result라는 suffix를 가져가는게 더 직관적일것 같은데 어때?.?

Hansangjin98 replied:
그게 더 훨씬 좋은 것 같다!!! 반영완료~!

nil-ian commented on Targets/Domain/Sources/UseCases/UploadVideoPostUseCase.swift:
```diff
@@ -0,0 +1,63 @@
+final class UploadVideoPostUseCase: UploadVideoPostUseCaseType {
+
+  // MARK: Properties
+  public let uploadVideoPostEvent: PublishSubject<UploadVideoPostEvent> = .init()
```
- 이 Subject는 다른 스트림과는 다르게 별도 suffix를 갖지 않고있는 이유가 있을까?

- 부모에서 자식으로의 이벤트는 rx스트림으로 연결하고 자식에서 부모로의 이벤트는 delegate로 연결하기로 했던것 같은데 확인 부탁쓰!

Hansangjin98 replied:
suffix 실수로 빠뜨렸다 ㅠㅠ 반영 완료!!

rx스트림으로 연결한 이유는 subject에서 값이 방출되었을 때 reactor에서 transform으로 mutation을 방출하려고 했던거였어!
delegate 방식으로도 구현할 수는 있어서 컨벤션, 가독성 측면에서 비교해보면 될 것 같은데 어떻게 생각해~?

nil-ian replied:
우리 앞전에 정했던 컨벤션은 자식에서 부모로의 이벤트는 delegate로 처리하기로 했어서 나는 그렇게 하고 있었어 ㅋㅋㅋ
이건 어느것이 더 나을지 좀 더 생각해봅시다~

nil-ian commented on Targets/Feature/Sources/VideoUpload/View/UploadVideoPostViewController.swift:
```diff
@@ -0,0 +1,388 @@
+    
+    reactor.state
+      .map { $0.isError }
+      .distinctUntilChanged()
```
error가 연달아 발생하게 되면 팝업은 한 번 뜨고 안뜰것 같은데 괜찮을까??

Hansangjin98 replied:
팝업창에서의 확인 버튼이 reactor action에 바인딩 되어있어서 isError값이 바뀌게 되는 구조라 괜찮을 것 같아!

nil-ian commented on Targets/Feature/Sources/VideoUpload/View/UploadVideoPostViewController.swift:
```diff
@@ -0,0 +1,388 @@
+      .filter { $0 }
+      .asSignal(onErrorSignalWith: .empty())
+      .emit(with: self, onNext: { owner, _ in
+        owner.present(owner.alertViewController, animated: true)
```
경고팝업도 Coordinator를 통하지 않고 VC에서 바로 띄우는 이유 궁금~

Hansangjin98 replied:
로딩VC랑 비슷한 이유가 있기도 했고, 추가적으로 팝업은 확인/취소 버튼에 대한 액션을 VC에서 작성해주고 initializer에 alertType을 넣어줘야해서 프로퍼티로 지녀야하는 상황이라 이왕 프로퍼티로 지니는 겸 코디네이터 스킵하고 VC에서 바로 띄웠어!

nil-ian commented on Targets/Feature/Sources/VideoUpload/View/UploadVideoPostViewController.swift:
```diff
@@ -0,0 +1,388 @@
+      .disposed(by: disposeBag)
+
+    reactor.state
+      .map { $0.hashTags.count <= 30 }
```
여기는 distintUntilChanged를 체이닝 하지 않은 기준이 어떤거야~?

Hansangjin98 replied:
hashTagErrorLabel의 text와 isHidden의 값을 변경하는 로직이 다른 곳에도 있는데
distintUntilChanged를 넣게 되면 text와 isHidden이 올바르게 바뀌지 않는 문제가 있어서 제외했어!

nil-ian commented on Targets/Feature/Sources/VideoUpload/View/UploadVideoPostViewController.swift:
```diff
@@ -0,0 +1,388 @@
+      .disposed(by: disposeBag)
+  }
+
+  func hashTagChecker(_ text: String) -> String {
```
1. 이 함수만 다른 함수들의 네이밍과는 다르게 명사형으로 네이밍한 이유가 있을까?
2. 해시태그의 어떤걸 체크하는지에 대한 의미도 네이밍에 같이 표현되면 좋을것 같아~

Hansangjin98 replied:
다시 보니 내가 봐도 별로인 네이밍이다 ㅜㅜ 반성하게 되네... 
변경 완료!

nil-ian commented on Targets/Service/Sources/Services/UploadVideoPostService.swift:
```diff
@@ -0,0 +1,75 @@
+              let videoFileURL: URL = .init(
+                string: PhoChakFileManager.shared.fetchVideoURLString(name: videoFile.fileName)
+              )
+        else { return .just("") }
```
presignedURL 혹은 videoFIleURL 언래핑 도중 에러로 42번, 45번 라인 guard else를 타게 되면 uploadKey에 빈 스트링이 들어간 형태로 API 호출을 하게 될 것 같은데 이 경우에는 서버와 협의해서 무효화하는 등 별도 예외처리를 해야 우리 크레딧을 좀 더 아낄 수 있을듯!

Hansangjin98 replied:
언래핑 도중에 에러가 발생하면 Single자체를 fail로 return 해서 서버 API를 안타도록 변경했는데 괜찮을까?

nil-ian reviewed (commented):
작업량이 엄청 많았네 ㅠㅠ 고생했어!!

Hansangjin98 committed:
feat: #15 - FileManager 관련 PR 리뷰 반영

Hansangjin98 committed:
feat: #15 - 로딩뷰 관련 PR 리뷰 반영

Hansangjin98 committed:
feat: #15 - 네이밍 관련 PR 리뷰 반영

Hansangjin98 committed:
feat: #15 - 예외처리 관련 PR 리뷰 반영

Hansangjin98 committed:
fix: #15 - Merge Conflict 수정

Hansangjin98 committed:
feat: #15 - FileManager 관련 PR 리뷰 반영

- FileManager가 필요한 곳은 주입을 받아 사용하는 것으로 변경
- saveVideo시 기존에 이미 동일 파일이 있는지 확인하는 로직 추가
- Coordinator 관련 오타 수정

Hansangjin98 committed:
Merge branch 'develop' into feature/upload
